### PR TITLE
[18.09] Keybase updates 

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
Thought it would be a good idea to backport as certain users were having problems, and it seems keybase has certain restrictions with which client versions can communicate with their servers.

cc @rvolosatovs 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by kalbasit** [comment]
I'm getting an error when I try to run keybase-gui:

```console
 pure  λ  nix-shell -p nix-review --run 'nix-review pr 55429'
 impure  λ  keybase service &
[1] 5415
▶ INFO | net.Listen on unix:/run/user/2000/keybase/keybased.sock
 impure  λ  NIX_SKIP_KEYBASE_CHECKS=1 keybase-gui
Could not find kbfsfuse client in keybase status.
You might need to run: kbfsfuse
/nix/store/n00sfd4bvlc3q4gckxa8a4v5d7cfa91r-keybase-gui-3.0.0/bin/.keybase-gui-wrapped: line 23: /nix/store/n00sfd4bvlc3q4gckxa8a4v5d7cfa91r-keybase-gui-3.0.0/share/keybase/Keybase: No such file or directory
 impure  λ  ls -la /nix/store/n00sfd4bvlc3q4gckxa8a4v5d7cfa91r-keybase-gui-3.0.0/share/keybase/Keybase
.r-xr-xr-x 116M root 31 Dec  1969 /nix/store/n00sfd4bvlc3q4gckxa8a4v5d7cfa91r-keybase-gui-3.0.0/share/keybase/Keybase
```

File: pkgs/tools/security/keybase/default.nix [hunk 1]
```diff
@@ -1,22 +1,20 @@
...
   subPackages = [ "go/keybase" ];
 
   dontRenameImports = true;
 
-  src = fetchFromGitHub {
-    owner  = "keybase";
-    repo   = "client";
-    rev    = "v${version}";
-    sha256 = "1gfxnqzs8msxmykg1zrhrrl2slmb29gl7b8s4m2g44zxaj91gfi9";
+  src = fetchurl {
```
Comment (rvolosatovs) [inline pkgs/tools/security/keybase/default.nix]: why this change?

Comment (kalbasit) [reply pkgs/tools/security/keybase/default.nix]: Using `fetchFromGitHub` does not work on Darwin. Please see 8da3c3648a0279f78a130abe10941a68fec7240a for context.

Comment (rvolosatovs) [reply pkgs/tools/security/keybase/default.nix]: Why not just redefine` fetchFromGitHub` as(simplified):
```nix
{
   fetchFromGitHub = {owner, repo, rev, sha256}: fetchurl {
      inherit sha256;
      url = "https://github.com/${owner}/${repo}/archive/${rev}.tar.gz"
   };
}
```
Then?
There are plenty of packages using the function(which AFAIK was considered the best practice), and so we'd need to change all invocations in `nixpkgs` to ensure consistency and compatibility with MacOS.

Comment (worldofpeace) [reply pkgs/tools/security/keybase/default.nix]: We'd get a major loss there as the automatically generated archives are not stable and can result in a different hash, which is something we don't want. It's less a problem to just use fetchurl for darwin and open that risk as this would never build on that platform if we didn't.

Commit msg for context
> fetchFromGitHub and thus fetchzip hashes the contents of the archive and
not the archive itself. Unicode file names lead to different checksums
on HFS+ vs. other file systems because of Unicode normalisation

Though it might be worth checking if we can eliminate this problem in the near future by identifying the problematic file upstream.

**Comment by worldofpeace** [comment]
Just got wind that our kbfs is out of date. See https://github.com/NixOS/nixpkgs/issues/50999#issuecomment-461852250

**Comment by worldofpeace** [comment]
@kalbasit I can reproduce the issue you're having with the 18.09 version.

On master works fine.

File: pkgs/tools/security/keybase/default.nix (hunk 1 above)
Comment (kalbasit) [reply pkgs/tools/security/keybase/default.nix]: > Though it might be worth checking if we can eliminate this problem in the near future by identifying the problematic file upstream.

I did not like this solution either, I would love to find a way for us to hash the contents regardless of the Filesystem. Keep in mind also that most Darwin machines are now sold pre-installed with APFS, which does not suffer from this issue.

**Comment by kalbasit** [comment]
@heronhaye we're having an interesting issue with the latest Keybase on release 18.09 (see [my comment above](https://github.com/NixOS/nixpkgs/pull/55429#issuecomment-461690389)). Do you know what could be causing this?

**Comment by worldofpeace** [comment]
Possible solution for `keybase-gui` is that we could stop patching the electron blob and just use electron_4 in nixpkgs.

**Comment by worldofpeace** [comment]
Ok, realized that it's obviously a problem with `autoPatchelfHook` and backporting all the changes to it fixes this and another backport pr that uses it.

Not sure if that could cause problems for stable, but I'd like to care for `18.09` as best I can until the next release.

**Comment by kalbasit** [comment]
@worldofpeace #55538 has been merged, you should probably rebase this PR so I can give it another try.

**Comment by worldofpeace** [comment]
@kalbasit Incoming in a moment 

I wasn't sure what would be the best state to have `18.09` in for keybase.
I guess having just kbfs outdated instead of everything would be better.
We should probably fix https://github.com/NixOS/nixpkgs/pull/55679 soon, but it's not a discussion for `18.09`

**Review by kalbasit — APPROVED** [review]

**Comment by kalbasit** [comment]
@worldofpeace it looks good now, thanks!

**Comment by worldofpeace** [comment]
You're welcome @kalbasit :sparkle:
